framework/src/mesh/FaceInfo.C

```diff
@@ -16,47 +16,70 @@
-                                      : std::numeric_limits<unsigned int>::max()),
-    _neighbor_centroid(_valid_neighbor ? neighbor->vertex_average()
-                                       : 2 * (_face_centroid - _elem_centroid) + _elem_centroid),
-    _neighbor_volume(_valid_neighbor ? neighbor->volume() : _elem_volume),
-    _d_cf(_neighbor_centroid - _elem_centroid),
-    _d_cf_mag(_d_cf.norm()),
-    _e_cf(_d_cf / _d_cf_mag)
+    _neighbor_side_id(std::numeric_limits<unsigned int>::max()),
+    _gc(0.5)
 {
-  // compute an centroid face normal by using 1-point quadrature
-  unsigned int dim = elem->dim();
-  std::unique_ptr<FEBase> fe(FEBase::build(dim, FEType(elem->default_order())));
-  QGauss qface(dim - 1, CONSTANT);
-  fe->attach_quadrature_rule(&qface);
-  const std::vector<Point> & normals = fe->get_normals();
-  fe->reinit(elem, _elem_side_id);
-  mooseAssert(normals.size() == 1, "FaceInfo construction broken w.r.t. computing face normals");
-  _normal = normals[0];
-
-  // Compute the position of the intersection of e_CF and the surface
-  _r_intersection =
-      _elem_centroid + (((_face_centroid - _elem_centroid) * _normal) / (_e_cf * _normal)) * _e_cf;
+  // Compute face-related quantities
+  const std::unique_ptr<const Elem> face = _elem_info->elem()->build_side_ptr(_elem_side_id);
+  _face_area = face->volume();
+  _face_centroid = face->vertex_average();
+
+  // Compute the face-normals
+  unsigned int dim = _elem_info->elem()->dim();
+  const auto r_cf = _face_centroid - _elem_info->centroid();
+
+  // For 1D elements, this is simple
+  if (dim == 1)
+    _normal = r_cf / r_cf.norm();
+  // For 2D elements, this is equally simple, we just need to make sure that
+  // the normal points in the right direction.
+  else if (dim == 2)
+  {
+    Point side = face->node_ref(0) - face->node_ref(1);
+    _normal = Point(-side(1), side(0));
+    _normal /= _normal.norm();
+    if (_normal * r_cf < 0.0)
+      _normal *= -1.0;
+  }
+  // In 3D we need to use the vector product
+  else
+  {
+    Point side_1 = face->node_ref(0) - face->node_ref(1);
+    Point side_2 = face->node_ref(0) - face->node_ref(2);
+    _normal = side_1.cross(side_2);
+    _normal /= _normal.norm();
+    if (_normal * r_cf < 0.0)
+      _normal *= -1.0;
+  }
+}
+
+void
+FaceInfo::computeCoefficients(const ElemInfo * const neighbor_info)
+{
+  mooseAssert(neighbor_info,
+              "We need a neighbor if we want to compute interpolation coefficients!");
+  _neighbor_info = neighbor_info;
+
+  if (_neighbor_info->isGhost())
+    _neighbor_side_id = std::numeric_limits<unsigned int>::max();
+  else
+    _neighbor_side_id = _neighbor_info->elem()->which_neighbor_am_i(_elem_info->elem());
+
+  // Setup quantities used for the approximation of the spatial derivatives
+  _d_cn = _neighbor_info->centroid() - _elem_info->centroid();
+  _d_cn_mag = _d_cn.norm();
+  _e_cn = _d_cn / _d_cn_mag;
+
+  Point r_intersection =
```

Comment:
```suggestion
  const auto r_intersection =
```
or `const Point` if you'd prefer to keep the explicit typing